### PR TITLE
Improve formatCurrency short form symbol formatting (with exceptions)

### LINKED ISSUE #2304: [react-i18n] Improve formatCurrency `short` form symbol formatting
## Overview

Related to #2265 and #1892 and https://github.com/Shopify/core-issues/issues/45203

Improve the `short` form of `formatCurrency` so that it more closely adheres to the [Polaris currency formatting guidelines](https://polaris.shopify.com/foundations/formatting-localized-currency), while also applying Shopify-defined exceptions for specific currencies where the longer form symbol should be displayed.

The `formatCurrencyShort` method calls `getShortCurrencySymbol`. This method calls `getCurrency`, which calls  [`Intl.NumberFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) with `style: 'currency'`.

If the formatted currency comes back with the first two characters of the currency and a symbol (e.g. `CA$`), the first two characters are stripped off (e.g. `$1.50` instead of `CA$1.50`).

However, we've been discovering some edge cases where a longer-form currency symbol is being returned when it shouldn't be:
- When formatting a `short` currency in `AUD` when the default is another currency, the currency symbol is `A$` instead of `$`.
- When formatting a `short` currency in `EGP`, the currency symbol is `EGP` instead of `E£`

#### Short currency symbol formatting improvement

By default , when `Intl.NumberFormat` has `style: 'currency'`, it uses the option `currencyDisplay: 'symbol'`.

However, the option `currencyDisplay: 'narrowSymbol'` might be more appropriate for more use cases, without the need to strip the first two characters of the symbol.

<details>
<summary>Example</summary>

```javascript
const number = 123456.789;

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'AUD', currencyDisplay: 'symbol'}).format(number));
// AUD symbol: "A$123,456.79"

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'AUD', currencyDisplay: 'narrowSymbol'}).format(number));
// AUD narrowSymbol: "$123,456.79"

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'CAD', currencyDisplay: 'symbol'}).format(number));
// CAD symbol: "CA$123,456.79"

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'CAD', currencyDisplay: 'narrowSymbol'}).format(number));
// CAD narrowSymbol: "$123,456.79"

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'EGP', currencyDisplay: 'symbol'}).format(number));
// EGP symbol: "EGP 123,456.79"

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'EGP', currencyDisplay: 'narrowSymbol'}).format(number));
// EGP narrowSymbol: "E£123,456.79"

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'CHF', currencyDisplay: 'symbol'}).format(number));
// CHF symbol: "CHF 123,456.79"

console.log(new Intl.NumberFormat('en', { style: 'currency', currency: 'CHF', currencyDisplay: 'narrowSymbol'}).format(number));
// CHF narrowSymbol: "CHF 123,456.79"
```
</details>

#### Exceptions

As mentioned in #2265, there are Shopify-defined exceptions for some currencies where the longer form symbol should be displayed.

As of right now, the exceptions should include `BRL` (Brazilian Real) and `HKD` (Hong Kong Dollar), which should show the full-length symbol across Shopify (`R$` and `HK$`, respectively).

#### Method Exposure

As mentioned in #1892, though we want to provide an opinionated `formatCurrency` library which makes it easy to format currencies according to Polaris guidelines, it would be useful to expose `getShortCurrencySymbol` and `getCurrencySymbol` publicly so that developers can retrieve the short currency symbol on its own and call `getCurrencySymbol` with `Intl.NumberFormat` options of their choosing.

#### Audit

As part of implementing this, I would want to generate output for `getShortCurrencySymbol` before and after the method change for all our currently supported currencies, for all our currently supported locales, and perform an audit on them by consulting with the [Common Locale Data Repository (CLDR)](https://cldr.unicode.org/) and internal Shopify currency experts.

## Type

- [x] New feature
- [x] Changes to existing features

## Motivation

Greater currency formatting consistency across currencies and locales and greater adherence to Polaris currency formatting guidelines.

## Checklist

- [x] Please delete the labels section before submitting your issue
- [x] I have described this issue in a way that is actionable (if possible)

### PR DESCRIPTION
**Pull request by sunical**
## Description

Fixes #2304
Part of https://github.com/Shopify/web/issues/70196

- Modifies `i18n.formatCurrency` `short` form to use `narrowSymbol` for `currencyDisplay` instead of the default `symbol`. For old browsers that don't support `narrowSymbol`, it defaults to `symbol`. The changes due to this modification are shown on [this spreadsheet](https://docs.google.com/spreadsheets/d/1NnAp1k-2MoBFNVpCJ9ttZyKQJAJO6C-rgtxR362bojQ/edit#gid=0), which lists all the supported currencies and the "before" and "after" outputs.
- Modifies `i18n.getCurrencySymbol` to return the short form of `i18n.formatCurrency`. It no longer displays a character or ISO code beside the currency symbol.
- Creates exceptions for BRL (Brazilian Real) and HKD (Hong Kong Dollar), which should show the full-length symbol across Shopify (R$ and HK$, respectively).

For auditing purposes, here's the [Spreadsheet](https://docs.google.com/spreadsheets/d/1NnAp1k-2MoBFNVpCJ9ttZyKQJAJO6C-rgtxR362bojQ/edit#gid=0) containing the "before" and "after" states of each supported currency given the modification to the `formatCurrency` `short` form.

### CODE REVIEW

**Review by cejaekl — APPROVED** [review]

File: packages/react-i18n/src/i18n.ts
Comment (BPScott) [inline packages/react-i18n/src/i18n.ts:495]: https://caniuse.com/?search=currencyDisplay suggests that `currencyDisplay: 'narrowSymbol'`is only supported in iOS 14.5 and greater. However the minimum version of iOS that [web currently supports is 12.2, and may be bumped to 13.0](https://github.com/Shopify/web/issues/75258), thanks to old point-of-sale devices rendering content from the web codebase in webviews.

I don't think we can use this without wrapping it in some kind of try/catch with a fallback for those old browsers.

File: packages/react-i18n/src/tests/i18n.test.ts
Comment (BPScott) [inline packages/react-i18n/src/tests/i18n.test.ts:2177]: Nitpick: For the sake of a more skimmable test, what do you think to using the array notation to avoid a few linebreaks?:

```
it.each([
  ['en', 'EUR', {}, {symbol: '€', prefixed: true}],
  ['fr', 'EUR', {}, {symbol: '€', prefixed: false}],
  ['en', 'SGD', {currencyDisplay: 'narrowSymbol'}, {symbol: 'SGD', prefixed: true}],
  ['en', 'SGD', {}, {symbol: '$', prefixed: true}],
])('returns symbol and position of currency ($currency), given the provided locale ($locale) and options ($options))',
  (locale, currency, options, expected) => {
```

**Review by BPScott — CHANGES_REQUESTED** [review]
I suspect `currencyDisplay: 'narrowSymbol'` usage is problematic given our support for old versions of ios (links to relevant tickets/version numbers inline).

We'll need to test in an old iOS version to see what happens when we ask for an unknown currencyDisplay value. I suspect it'll throw an error, and we'll need to either not use `narrowSymbol` or do try to use it, catch the error then try something else. Perhaps we need to stick with the default currencyDisplay, and have a longer exceptions list?

File: packages/react-i18n/src/i18n.ts
Comment (sunical) [reply packages/react-i18n/src/i18n.ts]: Will add testing for this now.

File: .changeset/quiet-foxes-attend.md
Comment (sunical) [reply .changeset/quiet-foxes-attend.md:5]: Ah, thanks.

**Comment by BPScott** [comment]
> Looks good! Before shipping, I recommend that you create a beta release (you can ask someone on Admin Web Platform to help with this).

Ohh that's easy. We have a new toy. Check this out:

File: packages/react-i18n/src/currencyShortFormException.ts
Comment (sunical) [reply packages/react-i18n/src/currencyShortFormException.ts:1]: I had this thought in my head on where to put this, and the `constants` folder was an option. But because `currencyCode.ts` was in the `src` folder I felt `currencyShortFormException.ts` could live there too. Now on a second thought, I think the `constants` folder would be more appropriate for it.

**Comment by sunical** [comment]
> /snapit

Now this is awesome!

File: packages/react-i18n/src/tests/i18n.test.ts
Comment (trishrempel) [reply packages/react-i18n/src/tests/i18n.test.ts:2168]: Oh right, I thought [this PR](https://github.com/Shopify/quilt/pull/2330) was shipped already 🤦🏻 👍🏻

**Review by BPScott — COMMENTED** [review]
Had a thought about the shape of the overrides inline.


I've just merged #2230 to move getShortCurrencySymbol from public to private, so you'll need to rebase. You should now be able to do `i18n.getShortCurrencySymbol` in the tests now.

File: .changeset/quiet-foxes-attend.md
Comment (BPScott) [reply .changeset/quiet-foxes-attend.md:2]: > I'd consider the new behaviour to be something everyone should opt into without thinking about.

To me that pushes this change more towards being a `minor` rather than `major` release - major would be "you need to think about and plan for using this behaviour as it isn't trivial to drop in" which is sounds like this isn't :). The last variable is "does this require much test churn", and hopefully the answer is "not much".

File: packages/react-i18n/src/i18n.ts
Comment (BPScott) [reply packages/react-i18n/src/i18n.ts]: I have no idea. I don't know what this value represents off the top of my head and I don't know if it changes per locale (a counter example would be fantastic showcasing prefix changing per locale). If it does change per locale then yeah keeping this as-is sounds better.

Comment (sunical) [reply packages/react-i18n/src/i18n.ts]: From my investigation, this changes depending on the `locale`. And this prefixed value dictates whether the symbol would `prefix` or `suffix` the amount value (see [here](https://github.com/Shopify/quilt/blob/15d9efb38274d06f6953944b1541be9ed44b5675/packages/react-i18n/src/i18n.ts#L466)).

Looking at the `getCurrencySymbol` [here](https://github.com/Shopify/quilt/blob/15d9efb38274d06f6953944b1541be9ed44b5675/packages/react-i18n/src/utilities/money.ts#L3), and then the currencyPrefix [here](https://github.com/Shopify/quilt/blob/15d9efb38274d06f6953944b1541be9ed44b5675/packages/react-i18n/src/utilities/money.ts#L23), this is where the prefixed value ultimately comes from and it seems reliant to the locale.

So I think we could leave it as is, otherwise, for each currency, we'll need to hardcode its `prefixed` value for each `locale`. Thanks!

**Review by BPScott — COMMENTED** [review]
Core code looks great! I'll defer to the translations team to confirm that the changes in the spreadsheet are desired and accurate. If 
they're happy I'm happy.

Those tests for checking behaviour when narrowSymbol throws are fantastic!

It looks like there's a few currencies mentioned in the spreadsheet that aren't mentioned in the quilt tests. Should we add those currencies to the unit tests so they're easier to check in the future?


I've got a few small thoughts inline, mostly around how we could polish documentation / changelog notes.

File: packages/react-i18n/src/i18n.ts
Comment (trishrempel) [inline packages/react-i18n/src/i18n.ts:347]: Something's been bugging me about the API design, and I haven't been able to think about why until now.

We want this API to be opinionated towards Shopify's Polaris guidelines on internationalization, and so `formatCurrency` is using `getShortCurrencySymbol` to get the short form currency.

However, we're exposing `getCurrencySymbol`, `getCurrencySymbolLocalized`, and `getShortCurrencySymbol`, and not making it clear that `getShortCurrencySymbol` is the recommended method.

I have a bit of a wild proposal:
- Update `getCurrencySymbolLocalized` to actually have the same logic as `getShortCurrencySymbol`, but allow the `options` parameter of type `Pick<Intl.NumberFormatOptions, 'currencyDisplay'>` to allow someone to override it if they want to.
- The `getCurrencySymbol` method remains as it is, also passing along the `options` parameter.
- We remove the `getShortCurrencySymbol`.

This would have an added benefit of automatically fixing the currency formatting for any code that is currently calling `getCurrencySymbol` or `getCurrencySymbolLocalized`.

This would end up ultimately negating the previous PR that just went out to expose `getShortCurrencySymbol`, but I think it might be better to update the methods we currently have than to add another option to the mix.

Thoughts?

**Review by trishrempel — CHANGES_REQUESTED** [review]

File: packages/react-i18n/src/i18n.ts
Comment (trishrempel) [reply packages/react-i18n/src/i18n.ts:349]: > Could we get away with the options being just Pick<Intl.NumberFormatOptions, 'currencyDisplay'>?

Yeah, this makes sense to me.

File: packages/react-i18n/src/i18n.ts
Comment (BPScott) [reply packages/react-i18n/src/i18n.ts:347]: > We want this API to be opinionated towards Shopify's Polaris guidelines on internationalization,

Strong agree on this!

> This would end up ultimately negating the previous PR that just went out to expose getShortCurrencySymbol

We've merged the PR but not yet released it. If we revert it then we have no breaking change. I agree with improving our current functions over exposing multiple slightly different ones.

Yesterday I was thinking about if it'd be possible to remove getCurrencySymbolLocalized and make its behaviour be an option of getCurrencySymbol but I'm not sure that's possible.

File: packages/react-i18n/src/i18n.ts
Comment (sunical) [reply packages/react-i18n/src/i18n.ts:365]: `options?: CurrencyFormatOptions` doesn't contain the `currency` property, so for now (except we modify it), it's not possible to pass `currency` via `options`, hence the separation.

File: packages/react-i18n/src/i18n.ts
Comment (sunical) [reply packages/react-i18n/src/i18n.ts:380]: Cool. We'll add it.

File: .changeset/quiet-foxes-attend.md
Comment (sunical) [reply .changeset/quiet-foxes-attend.md:5]: Yes, I think we can think of a better description here

File: packages/react-i18n/src/i18n.ts
Comment (sunical) [reply packages/react-i18n/src/i18n.ts:365]: Yea it extends it but when you check the definition via the repo, the `currency` property is missing in  `Intl.NumberFormatOptions`

**Review by BPScott — APPROVED** [review]
This looks fantastic! Code tests and changelog all look great to me!

As discussed, prior to merging, cut a snapshot release to test it within the `web` codebase. Once you've confirmed that behaviour is good with @trishrempel then merge.

Once you've merged give me a poke and I'll do the steps to release the new version of the package.

File: packages/react-i18n/src/i18n.ts
Comment (sunical) [reply packages/react-i18n/src/i18n.ts]: Finally confirmed the `currency` option exists in `Intl.NumberFormatOptions`

File: packages/react-i18n/src/i18n.ts
Comment (sunical) [reply packages/react-i18n/src/i18n.ts:346]: @trishrempel: A null check will still be needed somewhere: either in `getCurrencySymbol` or `getShortCurrencySymbol`. Otherwise, users will be able to call `formatCurrency` without entering a currency value. This is because `defaultCurrency` in the constructor is optional ([see here](https://github.com/Shopify/quilt/blob/f6ba8931b7ce598b3102e1c720c9ce47e7237776/packages/react-i18n/src/i18n.ts#L75)).

File: packages/react-i18n/src/i18n.ts
Comment (sunical) [reply packages/react-i18n/src/i18n.ts:486]: Yea I modified my comment to be clearer. `currency` is `string` so removing `||''` would lead to a `string = string | undefined` scenario which would return a TS warning.

If `this.defaultCurrency` is `undefined`, a casting would return a string `"undefined"` which is not what we want. I'll consider this resolved then.

**Review by trishrempel — APPROVED** [review]

File: .changeset/quiet-foxes-attend.md
Comment (trishrempel) [inline .changeset/quiet-foxes-attend.md:15]:
```suggestion
Deprecate usage of `i18n.getCurrencySymbolLocalized(locale, currency)`. Use `i18n.getCurrencySymbol(currency, locale)` instead.
```

**Pull request merged**
